warehouse/oidc/forms.py

```diff
@@ -0,0 +1,97 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import re
+
+import requests
+import wtforms
+
+from warehouse import forms
+from warehouse.i18n import localize as _
+
+# This roughly matches the "owner/repo" convention used by GitHub.
+_VALID_GITHUB_OWNER_REPO_SLUG = re.compile(
+    r"^[a-zA-Z0-9][a-zA-Z0-9-]*/[a-zA-Z0-9-_.]+$"
+)
+
+
+class GitHubProviderForm(forms.Form):
```

Comment:
Doesn't seem like we're checking anywhere if this particular provider configuration already exists. We should make sure to enforce uniqueness here.

Reply:
We have a uniqueness constraint on the DB, and I put the handling for that in the view logic itself (since the view is responsible for creating a model if it doesn't exist, and other views do similar things). But I can move it into the form validation layer, if you think that's more appropriate!

Reply:
https://github.com/pypa/warehouse/pull/10753/commits/1fc826a89e296274fcd57d2691b2d5aaba56ac73